include/aws/common/hash_table.h

```diff
@@ -76,6 +74,20 @@ struct aws_hash_element {
     void *value;
 };
 
+struct aws_hash_iter {
+    const struct aws_hash_table *map;
+    struct aws_hash_element element;
+    size_t slot;
+/*
+ * Reserving extra fields for binary compatibility with future expansion of
+ * iterator in case hash table implementation changes.
+ */
+    void * unused_0;
```

Comment:
So struct aws_hash_iter is designed to be declared locally where it is used and not take any memory allocations, (so it doesn't need to be freed afterward,) It needs to locally maintain state so that it keeps track of where in the hash table it is. For the current implementation, just a slot number is enough, but for other types of hash table implementations we might need other fields. I believe the idea here is to maintain ABI compatibility moving forward by making sure that we don't need to change the size of this struct later. So for instance, if a later implementation required the struct to use two size_t variables instead of one to maintain state, we would change the first of those (void *) elements to another size_t and still maintain the same overall size of the struct.

This was @bdonlan's idea, because he is often concerned about ABI compatibility of the SDK moving forward. I am not as concerned with this issue as he is but figured that it wasn't worth quibbling about. I hope that I have explained his reasoning on this properly, and hopefully he can chime in if he has something to add.